Add BatchBucketizeOp in caffe2

nateanl opened the pull request:
Summary: The operator transform dense features to sparse features by bucketizing. Only the feature in indices tensor will be transformed and output.

Differential Revision: D8820351

bddppq commented on caffe2/operators/batch_bucketize_op.cc:
```diff
@@ -0,0 +1,126 @@
+The float_features is a N * D tensor where N is the batch_size, and D is the feature_dim.
+The indices is a 1D tensor containing the indices of the features that need to be bucketized.
+The lengths is a 1D tensor that splits the following 'borders' argument.
+The borders is a 1D tensor containing the border list for each feature.
```
Could you use `boundaries` like in BatchBucketOneHot

bddppq commented on caffe2/operators/batch_bucketize_op.cc:
```diff
@@ -0,0 +1,126 @@
+the number of borders in the sub border list.
+The sum of all elements in `lengths` must be equal to the size of  `borders`.
+If lengths[0] = 2, the first sub border list is [0.5, 1.0], which separate the
+value to (-inf, 0.5), [0,5, 1.0), [1.0, inf). The bucketized feature will have
```
Could you do right-inclusive like in `BatchBucketOneHot`

nateanl replied:
Will change that.

bddppq commented on caffe2/operators/batch_bucketize_op.cc:
```diff
@@ -0,0 +1,126 @@
+    .SetDoc(R"DOC(
+Bucketize the float_features into sparse features.
+The float_features is a N * D tensor where N is the batch_size, and D is the feature_dim.
+The indices is a 1D tensor containing the indices of the features that need to be bucketized.
```
I think you can first do Gather/BatchGather instead of adding this `indices` argument into `BatchBucketize`

nateanl replied:
Will change that.

nateanl replied:
I think it'll be better to bucketize features by one operator. Maybe we can set up a meeting tomorrow to talk about this.

bddppq replied:
We in general discourage fused operators (unless it's for performance reason in critical operators, like `bias` in Convolution), because it's in general less expressive.

amirshim replied:
@bddppq The extra code here for a fused operation is minimal, and it doesn't make the operator any less expressive. On the other hand, When you're talking about large amounts of data flowing through for preprocessing features, it reduces copying the data by 2, most likely making the operator almost 2 times faster. (as opposed to using BatchGather first and then this operator)

bddppq replied:
@amirshim In your use case, normally what's the size of the batched features matrix?

bddppq replied:
chatted offline, fusion is a right compromise here.

bddppq commented on caffe2/operators/batch_bucketize_op.cc:
```diff
@@ -0,0 +1,126 @@
+        2,
+        "borders",
+        "Flatten tensor, dimension has to match the sum of lengths")
+    .Input(
```
convention is to put data as the first input

bddppq commented on caffe2/operators/batch_bucketize_op.cc:
```diff
@@ -0,0 +1,126 @@
+    .Input(
+        1,
+        "lengths",
+        "Flatten tensor, which the size must be equal to indices")
```
could you document that the elem type should be int32?

bddppq commented on caffe2/python/operator_test/batch_bucketize_op_test.py:
```diff
@@ -0,0 +1,91 @@
+        workspace.RunOperatorOnce(core.CreateOperator('BatchBucketize',
+                                 ["INDICES", "BORDERS", "LENGTHS", "FEATURE"], ["O"]))
+
+        assert((workspace.FetchBlob('O') == ref(x, lens, boundaries, indices)).all())
```
there is `self.assertReferenceChecks`

nateanl committed:
Add BatchBucketizeOp in caffe2 (#9385)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/9385

The operator transform dense features to sparse features by bucketizing. Only the feature in indices tensor will be transformed and output.

Differential Revision: D8820351

fbshipit-source-id: 7b23187713a2d370df5c1f9e4990c480c29b8969

bddppq reviewed (approved):
LGTM
